rav1e_js/src/lib.rs

```diff
@@ -7,88 +7,172 @@
-        break;
-      }
-      Err(EncoderStatus::NeedMoreData) => {
-        // The encoder has requested additional frames. Push the next frame
-        // in, or flush the encoder if there are no frames left (on None).
-        ctx.flush();
-      }
-      Err(e) => {
-        panic!(e);
-      }
+
+  /// Flushes the encoder.
+  ///
+  /// Flushing signals the end of the video. After the encoder has been flushed, no additional frames are accepted.
+  pub fn flush(&mut self) {
+    self.ctx.flush();
+  }
+
+  /// Encodes the next frame and returns the encoded data.
+  ///
+  /// This method is where the main encoding work is done.
+  #[allow(non_snake_case)]
+  pub fn encodeNext(&mut self) -> Result<Packet, JsValue> {
+    match self.ctx.receive_packet() {
+      Ok(packet) => Ok(Packet { p: packet }),
+      Err(e) => Err(construct_js_err(e, "")),
     }
   }
-  format!("Done encoding the same tiny blank frame {} times.", repeats)
+}
+
+/// Wrapper around the encoder configuration (`rav1e::api::config::encoder::EncoderConfig`).
+///
+/// Encoder settings which impact the produced bitstream.
+#[wasm_bindgen]
+#[derive(Clone)]
+pub struct EncoderConfig {
+  conf: Rav1eEncoderConfig,
+}
+
+#[wasm_bindgen]
+impl EncoderConfig {
+  #[wasm_bindgen(constructor)]
+  pub fn new() -> Self {
+    EncoderConfig { conf: Rav1eEncoderConfig::default() }
+  }
+
+  pub fn debug(&self) -> String {
+    format!("{:?}", self.conf)
+  }
+
+  /// Set width and height of the frames in pixels.
+  pub fn set_dim(&mut self, width: usize, height: usize) -> Self {
+    self.conf.height = height;
+    self.conf.width = width;
+    self.clone()
+  }
+
+  /// Set the speed setting according to a numeric speed preset.
+  ///
+  /// The speed settings vary depending on speed value from 0 to 10.
+  /// - 10 (fastest): min block size 64x64, reduced TX set, fast deblock, fast scenechange detection.
+  /// - 9: min block size 32x32, reduced TX set, fast deblock.
+  /// - 8: min block size 8x8, reduced TX set, fast deblock.
+  /// - 7: min block size 8x8, reduced TX set.
+  /// - 6 (default): min block size 8x8, reduced TX set, complex pred modes for keyframes.
+  /// - 5: min block size 8x8, complex pred modes for keyframes, RDO TX decision.
+  /// - 4: min block size 8x8, complex pred modes for keyframes, RDO TX decision, full SGR search.
+  /// - 3: min block size 8x8, complex pred modes for keyframes, RDO TX decision, include near MVs,
+  ///        full SGR search.
+  /// - 2: min block size 4x4, complex pred modes, RDO TX decision, include near MVs,
+  ///        full SGR search, coarse directions.
+  /// - 1: min block size 4x4, complex pred modes, RDO TX decision, include near MVs,
+  ///        bottom-up encoding, full SGR search.
+  /// - 0 (slowest): min block size 4x4, complex pred modes, RDO TX decision, include near MVs,
+  ///        bottom-up encoding with non-square partitions everywhere, full SGR search.
+  pub fn set_speed(mut self, speed: usize) -> Self {
```

Comment:
you may just return self if you take ownership of it. Make sure you are consistent though, set_dim is taking a mutable reference instead.

Reply:
The returning of `Self` is to enable method trailing

Reply:
And it should always be `&mut self`